File: challenges/partitionLinkedList/Solutions/PartitionLinkedList.java
package wwc;

/**
 * Given a linked list and a value x, partition it such that all nodes less than x come before nodes greater than or equal to x.
 * 
 *For example,
 *Given 1->4->3->2->5->2 and x = 3,
 *return 1->2->2->4->3->5. 
 *
 *@author Cintia F. Kasziba
 */
public class PartitionLinkedList {
	
	public static void main(String... args) {
		ListNode head = createList(1,4,3,2,5,2);
		printList(head);
		printList(partition(head, 3));
	}
	
	private static ListNode partition(ListNode head, int x) {
		
		ListNode lessThanList = null;
		ListNode greatherThanEqualList = null;
		ListNode cursor = head;
		ListNode lessThanTail = null;
		ListNode greatherThanTail = null;
		ListNode greatherThanEqualListHead = null;
		
		
		while(cursor != null) {
			if(cursor != null && cursor.val < x) {
				
				if(lessThanList == null) {
					lessThanList = cursor;
					lessThanTail = cursor;
				} else {
					lessThanTail.next = cursor;
					lessThanTail = cursor;
				}
				cursor = cursor.next;
				while(cursor != null && cursor.val < x) {
					lessThanTail = cursor;
					cursor = cursor.next;
				}
			}
			
			if(cursor != null && cursor.val >= x) {
				
				if(greatherThanEqualList == null) {
					greatherThanEqualList = cursor;
					greatherThanEqualListHead = cursor;
					greatherThanTail = cursor;
				} else {
					greatherThanTail.next = cursor;
					greatherThanTail = cursor;
				}
				cursor = cursor.next;
				while(cursor != null && cursor.val >= x) {
					greatherThanTail = cursor;
					cursor = cursor.next;
				}
			}
		}
		
		if(lessThanTail != null) {
			lessThanTail.next = greatherThanEqualListHead;
		}
		
		if(greatherThanEqualList != null) {
			greatherThanTail.next = null;
		}
		return head;
	}
	
	private static ListNode createList(int... elements) {
		if(elements.length == 0) {
			return null;
		}
		ListNode head = new ListNode(elements[0]);
		ListNode first = head;
		for(int i = 1; i < elements.length; i++) {
			head.next = new ListNode(elements[i]);
			head = head.next;
		}
		return first;
	}
	
	private static void printList(ListNode head) {
		ListNode cursor = head;
		while(cursor != null) {
			System.out.print(cursor.val + ", ");
			cursor = cursor.next;
		}
		System.out.println("");
	}
	
	private static class ListNode {
		int val;
		ListNode next;
		
		ListNode(int val) {
			this.val = val;
		}
	}
}

File: challenges/partitionLinkedList/Solutions/partitionLinkedList.cpp
#include <iostream>
#include <string>

using namespace std;
/*
WomenWhoCodeNYC July 2018 challenges
partition list
nodes smaller than 3 come before nodes greater of equal to 3 original relative order is prserved

test cases:
3
6
1 4 3 2 5 2

output:
1 2 2 4 3 5

try when first element is greater or equal than 3
3
6
4 1 3 2 5 2

output:
1 2 2 4 3 5
*/
class SinglyLinkedListNode {
public:
	int data;
	SinglyLinkedListNode *next;

	SinglyLinkedListNode(int node_data) {
		this->data = node_data;
		this->next = nullptr;
	}
};

class SinglyLinkedList {
public:
	SinglyLinkedListNode * head;
	SinglyLinkedListNode *tail;

	SinglyLinkedList() {
		this->head = nullptr;
		this->tail = nullptr;
	}

	void insert_node(int node_data) {
		SinglyLinkedListNode* node = new SinglyLinkedListNode(node_data);

		if (!this->head) {
			this->head = node;
		}
		else {
			this->tail->next = node;
		}

		this->tail = node;
	}

	/*
	keep track of begining and end of 2nd half 
	*/
	SinglyLinkedListNode * partition(SinglyLinkedListNode *head, int x)
	{
		SinglyLinkedListNode *it1 = nullptr, *it2 = nullptr;
		SinglyLinkedListNode *it1_prev = nullptr, *it2_prev = nullptr;
		SinglyLinkedListNode *node = head;
		int it1dist = 0, it2dist = 0;
		int itdist = 1;
		bool first_it2 = false;
		while (node != nullptr)
		{
			if (node->data < x)
			{
				it1_prev = it1;
				it1 = node;
				it1dist = itdist;
			}
			else
			{
				if (first_it2 == false)
				{
					it2_prev = node;
					it2 = node;
					it2dist = itdist;
					first_it2 = true;
				}
				else
				{
					it2 = node;
				}
			}

			if ((it1dist>0) && (it2dist>0) && (it1dist > it2dist))
			{// current element is out of place so needs to move backwards
				it2->next = it1->next;
				if (it1_prev)
					it1_prev->next = it1;
				else				
					head = it1;
				
				it1->next = it2_prev;
				 
				it1dist = it2dist;
				it2dist++;

				node = it2->next;
			}
			else
				node = node->next;
			itdist++;
		}
		return head;
	}
	
};


void print_singly_linked_list(SinglyLinkedListNode* node, string sep, ostream& fout) {
	while (node) {
		fout << node->data;

		node = node->next;

		if (node) {
			fout << sep;
		}
	}
}

void free_singly_linked_list(SinglyLinkedListNode* node) {
	while (node) {
		SinglyLinkedListNode* temp = node;
		node = node->next;

		free(temp);
	}
}

int main()
{
	int partition_val;
	cin >> partition_val;

	SinglyLinkedList* llist = new SinglyLinkedList();
	int llist_count;
	cin >> llist_count;
	for (int i = 0; i < llist_count; i++) {
		int llist_item;
		cin >> llist_item;
		llist->insert_node(llist_item);
	}
	SinglyLinkedListNode *head =  llist->partition(llist->head, partition_val);
	print_singly_linked_list(head, " ", cout);

	if (llist)
		free_singly_linked_list(llist->head);
}


